pkg/retriever/cidretrieval.go

```diff
@@ -0,0 +1,502 @@
+func (retrieval *cidRetrieval) RetrieveCid(ctx context.Context, cid cid.Cid) (*RetrievalStats, error) {
+	// state local to this CID's retrieval
+	state := &retrievalState{
+		Cid:               cid,
+		ResultChan:        make(chan runResult),
+		FinishChan:        make(chan struct{}),
+		WaitQueue:         prioritywaitqueue.New(queryCompare),
+		ActiveQueryCancel: make([]*func(), 0),
+	}
+
+	ctx, cancelCtx := context.WithCancel(ctx)
+	defer cancelCtx()
+
+	// fetch indexer candidates for CID
+	candidates, err := retrieval.findCandidates(ctx, cid)
+	if err != nil {
+		return nil, err
+	}
+
+	// start retrievals
+	retrieval.WaitGroup.Add(len(candidates))
+	for _, candidate := range candidates {
+		candidate := candidate
+		go func() {
+			retrieval.runRetrieval(ctx, state, candidate)
+			retrieval.WaitGroup.Done()
+		}()
+	}
+
+	return retrieval.collectResults(ctx, state, candidates)
+}
+
+// findCandidates calls the indexer for the given CID
+func (retrieval *cidRetrieval) findCandidates(ctx context.Context, cid cid.Cid) ([]RetrievalCandidate, error) {
+	candidates, err := retrieval.IndexEndpoint.FindCandidates(ctx, cid)
+	if err != nil {
+		return nil, fmt.Errorf("could not get retrieval candidates for %s: %w", cid, err)
+	}
+
+	retrieval.Instrumentation.OnRetrievalCandidatesFound(len(candidates))
+
+	if len(candidates) == 0 {
+		return nil, ErrNoCandidates
+	}
+
+	acceptableCandidates := make([]RetrievalCandidate, 0)
+	for _, candidate := range candidates {
+		if retrieval.IsAcceptableStorageProvider(candidate.MinerPeer.ID) {
+			acceptableCandidates = append(acceptableCandidates, candidate)
+		}
+	}
+
+	if err := retrieval.Instrumentation.OnRetrievalCandidatesFiltered(len(acceptableCandidates)); err != nil {
+		return nil, err
+	}
+
+	if len(acceptableCandidates) == 0 {
+		return nil, ErrNoCandidates
+	}
+
+	return acceptableCandidates, nil
+}
+
+// collectResults is responsible for receiving query errors, retrieval errors
+// and retrieval results and aggregating into an appropriate return of either
+// a complete RetrievalStats or an bundled multi-error
+func (retrieval *cidRetrieval) collectResults(ctx context.Context, state *retrievalState, candidates []RetrievalCandidate) (*RetrievalStats, error) {
+	var finishedCount int
+	var queryErrors error
+	var retrievalErrors error
+	var stats *RetrievalStats
+
+collectcomplete:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, context.Canceled
+		case result := <-state.ResultChan:
+			if result.QueryError != nil {
+				queryErrors = multierr.Append(queryErrors, result.QueryError)
+			}
+			if result.RetrievalError != nil {
+				retrievalErrors = multierr.Append(retrievalErrors, result.RetrievalError)
+			}
+			if result.RetrievalResult != nil {
+				stats = result.RetrievalResult
+				break collectcomplete
+			}
+			// have we got all responses but no success?
+			finishedCount++
+			if finishedCount >= len(candidates) {
+				break collectcomplete
```

Comment:
Is this so that it breaks out of the loop and not the select?

Reply:
yeah .. so the original version of this was a `for` across the channel, which was really clean, but then I realised I needed to be able to handle context cancellation, and I don't believe there's a good way I can do that, aside perhaps from closing the channel but that has other consequences _and_ that channel needs to be closed from some other goroutine! .. so this is it, and it's ugly, and it's the second block of code like this in this file now :shrug:.